tests/distributions/test_iaf.py

```diff
@@ -96,6 +97,64 @@ def nonzero(x):
 
             assert lower_sum == float(0.0)
 
+    def _test_masks(self, input_dim, observed_dim, hidden_dim, num_layers, permutation, output_dim_multiplier):
+        masks, mask_skip = create_mask(input_dim, observed_dim, hidden_dim,
+                                       num_layers, permutation, output_dim_multiplier)
+
+        # First test that hidden layer masks are adequately connected
+        # Tracing backwards, works out what inputs each output is connected to
+        # It's a dictionary of sets indexed by a tuple (input_dim, param_dim)
+        permutation = list(permutation.numpy())
+
+        # Loop over variables
+        for idx in range(input_dim):
+            # Calculate correct answer
+            correct = torch.cat((torch.arange(observed_dim), torch.tensor(
+                sorted(permutation[0:permutation.index(idx)]), dtype=torch.long) + observed_dim))
+
+            # Loop over parameters for each variable
+            for jdx in range(output_dim_multiplier):
+                prev_connections = set()
+                # Do output-to-penultimate hidden layer mask
+                for kdx in range(masks[-1].size(1)):
+                    if masks[-1][idx + jdx * input_dim, kdx]:
+                        prev_connections.add(kdx)
+
+                # Do hidden-to-hidden, and hidden-to-input layer masks
+                for m in reversed(masks[:-1]):
+                    this_connections = set()
+                    for kdx in prev_connections:
+                        for ldx in range(m.size(1)):
+                            if m[kdx, ldx]:
+                                this_connections.add(ldx)
+                    prev_connections = this_connections
+
+                assert (torch.tensor(list(sorted(prev_connections)), dtype=torch.long) == correct).all()
+
+                # Test the skip-connections mask
+                skip_connections = set()
+                for kdx in range(mask_skip.size(1)):
+                    if mask_skip[idx + jdx * input_dim, kdx]:
+                        skip_connections.add(kdx)
+                assert (torch.tensor(list(sorted(skip_connections)), dtype=torch.long) == correct).all()
+
     def test_jacobians(self):
         for input_dim in [2, 3, 5, 7, 9, 11]:
             self._test_jacobian(input_dim, 3 * input_dim + 1, 2)
+
+    def test_masks(self):
+        for input_dim in [1, 3, 5]:
+            for observed_dim in [0, 3]:
+                for num_layers in [1, 3]:
+                    for output_dim_multiplier in [1, 2, 3]:
+                        # NOTE: the hidden dimension must be greater than the input_dim for the
```

Comment:
thanks, these mask tests look great. 

"the hidden dimension must be greater..." is this checked in autoregressivenn somewhere?

Reply:
No but I've just pushed another commit that checks this. Also I added a comment about the mask structure